x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/retriever/PinnedRetrieverBuilder.java

```
@@ -0,0 +1,245 @@
+        }
+    }
+
+    public PinnedRetrieverBuilder(List<String> ids, List<SpecifiedDocument> docs, RetrieverBuilder retrieverBuilder, int rankWindowSize) {
+        super(new ArrayList<>(), rankWindowSize);
+        validateIdsAndDocs(ids, docs);
+        this.ids = ids != null ? ids : new ArrayList<>();
+        this.docs = docs != null ? docs : new ArrayList<>();
+        addChild(new PinnedRetrieverBuilderWrapper(retrieverBuilder));
+    }
+
+    public PinnedRetrieverBuilder(
+        List<String> ids,
+        List<SpecifiedDocument> docs,
+        List<RetrieverSource> retrieverSource,
+        int rankWindowSize,
+        String retrieverName,
+        List<QueryBuilder> preFilterQueryBuilders
+    ) {
+        super(retrieverSource, rankWindowSize);
+        validateIdsAndDocs(ids, docs);
+        this.ids = ids;
+        this.docs = docs;
+        this.retrieverName = retrieverName;
+        this.preFilterQueryBuilders = preFilterQueryBuilders;
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
+    public int rankWindowSize() {
+        return rankWindowSize;
+    }
+
+    /**
+     * Creates a PinnedQueryBuilder with the appropriate pinned documents.
+     * Prioritizes docs over ids if both are present.
+     *
+     * @param baseQuery the base query to pin documents to
+     * @return a PinnedQueryBuilder or the original query if no pinned documents
+     */
+    private QueryBuilder createPinnedQuery(QueryBuilder baseQuery) {
+        if (baseQuery == null) {
+            baseQuery = new MatchAllQueryBuilder();
+        }
+        if (docs.isEmpty() == false) {
+            return new PinnedQueryBuilder(baseQuery, docs.toArray(new SpecifiedDocument[0]));
+        } else if (ids.isEmpty() == false) {
+            return new PinnedQueryBuilder(baseQuery, ids.toArray(new String[0]));
+        } else {
+            return baseQuery;
```

This state should return an error, because no docs or ids were specified. Please validate this in the constructor so there's no need to handle this case.